mantidproject/mantid
# Display fit results in grouped workspaces and update model in custom setup

**Description of work.**

Every time a fit is successful the model string will be stored as a `custom setup' in the fit property browser under the name of the workspace fitted. This allows the function to be easily applied to other workspaces.
In addition there is a grouping of workspaces created that stores the fit value and error of parameters of the model for each loaded workspace that reproduces as closely as possible the tables in the hdf output of the legacy software. This aggregates the parameters of repeated functions (typically peaks) such that e.g. if three peaks are fitted then the tables of width, height etc. will have three column corresponding to the each of the three peaks. In order to reduce the number of workspaces the fit results for each parameter are stored in matrix worksapces where X corresponds to the index of the repeated function/peak (i.e. 1-3 for the above example), Y and E correspond to the fit value and error on a given parameter. The spectra correspond to the runs loaded (see the table of the UI fitting tab) - in general different models/functions could be fitted to each run, so when there is a parameter that does not exist for a run, the Y and E fields in the relevant row are filled with NaNs. There is also a table workspace that stores the model string that can be used to load the model into the fitpropertybrowser (Setup > Manage Setup > Load From String).

Deleting or renaming the data workspaces will result in the UI table being refreshed and could result in the rows of the fit tables not matching up to the rows un the logs and UI table. However once a subsequent fit has been successful then the fit table will be refreshed and the rows will match again. I can change this behaviour but I thought it's a bit of an edge case.

**To test:**
(1) Get data to fit by following instructions (1-2) in PR #28393 or downloading the data in PR #29130
(2) Load dat as in PRs above and add some workspaces to the plot (check plot box in the table)
(3) Toggle the fit property browser using the 'Fit' button on the toolbar.
(4) Add some functions (right-click to add peaks etc. as usual) and fit
(5) Check that there is a model stored under the fitted workspace name (Setup > Custom Setup > ws_name)
(6) Check the fit group of workspaces contains a workspace per parameter (naming convention FunctionName_ParameterName)
(7) Check that there is a row per run loaded in the table and that the fit values/errors corresponding to the row of the run fitted are correct
(8) Try fitting another workspace with different functions to the first and check the results as above. If a function parameter doesn't exist in the model of a given run, or the run hasn't been fitted at all, then the Y/E values should be NaN (in the latter instance the model string in the table workspace will be empty).

Partially fixes #27780

Don't merge until PR #29594 is merged

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
